### PR TITLE
Avoid too many copies in string encoder/decoder

### PR DESCRIPTION
**Pull request by evt**
Closes [#12](https://github.com/spacemeshos/go-scale/issues/12)

### CODE REVIEW

**Comment by fasmat** [comment]
@dshulyak I got the following results running the benchmarks in `unsafe_test.go`:

```
go test -v -benchmem -bench="^Benchmark_" -run="^$" -gcflags=-N .

goos: linux
goarch: arm64
pkg: github.com/spacemeshos/go-scale
Benchmark_UnsafeBytesToString
Benchmark_UnsafeBytesToString-6         530076894                2.265 ns/op           0 B/op          0 allocs/op
Benchmark_SafeBytesToString
Benchmark_SafeBytesToString-6           417205314                2.864 ns/op           0 B/op          0 allocs/op
Benchmark_UnsafeStringToBytes
Benchmark_UnsafeStringToBytes-6         348893032                3.448 ns/op           0 B/op          0 allocs/op
Benchmark_SafeStringToBytes
Benchmark_SafeStringToBytes-6           419229655                2.861 ns/op           0 B/op          0 allocs/op
PASS
ok      github.com/spacemeshos/go-scale 5.968s
```

I'm not sure if we should use the unsafe variants avoiding additional allocations, the go runtime will try to avoid allocations anyway if it can. In the case of Bytes2String we are ~ 30% faster, but in the case of String2Bytes we are ~ 20% slower than the safe way to do it.

What is your opinion @dshulyak .

**Comment by fasmat** [comment]
Using a `io.StringWriter` when possible improves encoding slightly when using optimizations:

```
go test -v -benchmem -bench="^BenchmarkEnc" -run="^$" .
goos: linux
goarch: arm64
pkg: github.com/spacemeshos/go-scale
BenchmarkEncodeStrings_WithStringWriter
BenchmarkEncodeStrings_WithStringWriter-6               39392080                32.31 ns/op           74 B/op          0 allocs/op
BenchmarkEncodeStrings_WithWriterForStrings
BenchmarkEncodeStrings_WithWriterForStrings-6           53234829                51.74 ns/op           40 B/op          0 allocs/op
PASS
ok      github.com/spacemeshos/go-scale 7.676s
```

without optimization there is almost no difference:

```
go test -v -benchmem -bench="^BenchmarkEnc" -run="^$" -gcflags=-N .
goos: linux
goarch: arm64
pkg: github.com/spacemeshos/go-scale
BenchmarkEncodeStrings_WithStringWriter
BenchmarkEncodeStrings_WithStringWriter-6               24001058                56.67 ns/op           62 B/op          0 allocs/op
BenchmarkEncodeStrings_WithWriterForStrings
BenchmarkEncodeStrings_WithWriterForStrings-6           25614274                54.75 ns/op           41 B/op          0 allocs/op
PASS
ok      github.com/spacemeshos/go-scale 6.168s
```

**Comment by fasmat** [comment]
I updated the benchmarks in a way that hopefully prevents go from optimizing the allocation away. These are the new results:

```
go test -v -benchmem -bench="^Benchmark_" -run="^$" -gcflags=-N .

goos: linux
goarch: arm64
pkg: github.com/spacemeshos/go-scale
Benchmark_UnsafeBytesToString
Benchmark_UnsafeBytesToString-6         539234050                2.224 ns/op           0 B/op          0 allocs/op
Benchmark_SafeBytesToString
Benchmark_SafeBytesToString-6           98168997                12.29 ns/op           16 B/op          1 allocs/op
Benchmark_UnsafeStringToBytes
Benchmark_UnsafeStringToBytes-6         350045876                3.431 ns/op           0 B/op          0 allocs/op
Benchmark_SafeStringToBytes
Benchmark_SafeStringToBytes-6           70134423                14.63 ns/op           16 B/op          1 allocs/op
PASS
ok      github.com/spacemeshos/go-scale 6.225s
```

```
go test -v -benchmem -bench="^BenchmarkEnc" -run="^$" -gcflags=-N .

goos: linux
goarch: arm64
pkg: github.com/spacemeshos/go-scale
BenchmarkEncodeStrings_WithStringWriter
BenchmarkEncodeStrings_WithStringWriter-6                4509775               312.8 ns/op            69 B/op          0 allocs/op
BenchmarkEncodeStrings_WithWriterForStrings
BenchmarkEncodeStrings_WithWriterForStrings-6            7346796               156.6 ns/op            36 B/op          0 allocs/op
PASS
ok      github.com/spacemeshos/go-scale 4.976s
```

**Review by dshulyak — APPROVED** [review]

File: unsafe.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package scale
+
+import (
+	"reflect"
+	"unsafe"
+)
+
+func StringToBytes(s string) []byte {
```
Comment (dshulyak) [inline unsafe.go:8]: this doesn't need to be public, right?

File: encoder_test.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	for i := 0; i < b.N; i++ {
+		EncodeString(enc, "Hello World")
+	}
+	b.StopTimer()
+
+	b.Log(buf.Len())
+}
+
+func BenchmarkEncodeStrings_WithWriterForStrings(b *testing.B) {
+	// bytes.Buffer does not implement the io.StringWriter interface.
```
Comment (dshulyak) [inline encoder_test.go:24]: but bytes.Buffer implements io.StringWriter :)

Comment (fasmat) [reply encoder_test.go:24]: Hm... Then I have to do a trick to measure the difference correctly 😄

**Pull request merged**